src/mame/drivers/dfs500.cpp

```diff
@@ -0,0 +1,1036 @@
+{
+	// PD71059C (Programmable Interrupt Controller)
+	return m_pic->read((offset>>1) & 1); // addressed by CPU's address bit AB1
+}
+
+void dfs500_state::pic_w(offs_t offset, uint8_t data)
+{
+	// PD71059C (Programmable Interrupt Controller)
+	m_pic->write((offset>>1) & 1, data); // addressed by CPU's address bit AB1
+}
+
+void dfs500_state::rombank1_entry_w(offs_t offset, uint8_t data)
+{
+	m_rombank1->set_entry(((data >> 1) & 0x40) | (data & 0x3f));
+}
+
+void dfs500_state::rombank2_entry_w(offs_t offset, uint8_t data)
+{
+	m_rombank2->set_entry(((data >> 1) & 0x40) | (data & 0x3f));
+}
+
+void dfs500_state::input_select_w(offs_t offset, uint8_t data)
+{
+	// Selects sources of video input on the AD-76 board.
+	m_input_sel_A = (data >> 3) & 0x7;
+	m_input_sel_B = data & 0x7;
+}
+
+void dfs500_state::WA0_w(offs_t offset, uint16_t data)
+{
+	m_effectcpu_latch16 = data;
+	m_TOC = true;
+}
+
+uint16_t dfs500_state::RA0_r(offs_t offset)
+{
+	return m_maincpu_latch16;
+	m_TOE = false;
+}
+
+uint16_t dfs500_state::RA1_r(offs_t offset)
+{
+	// "TEST, 1, OPT2, OPT1, RFLD, VD, TOC, TOE"
+	uint8_t value = 0x40;
+	// FIXME! Add other signals.
+	if (m_TOC) value |= 0x02;
+	if (m_TOE) value |= 0x01;
+	value |= ((ioport("DSW_S3")->read() & 0x0f) << 8); // (Unknown)
+	return value;
+}
+
+void dfs500_state::WB0_w(offs_t offset, uint16_t data)
+{
+	m_maincpu_latch16 = data;
+	m_TOE = true;
+}
+
+uint16_t dfs500_state::RB0_r(offs_t offset)
+{
+	return m_effectcpu_latch16;
+	m_TOC = false;
+}
+
+uint8_t dfs500_state::RB1_r(offs_t offset)
+{
+	//"TEST, OPT2, OPT1, VD, T2, T1, TOE, TOC"
+	uint8_t value = 0;
+	// FIXME! Add other signals.
+	if (m_TOE) value |= 0x02;
+	if (m_TOC) value |= 0x01;
+	return value;
+}
+
+uint8_t dfs500_state::RB2_r(offs_t offset)
+{
+	uint8_t value = 0;
+	value |= ((ioport("DSW_S1")->read() & 0x0f) << 4); // (Editing Control Unit Select)
+	value |= (ioport("DSW_S2")->read() & 0x0f);        // (Freeze Timing)
+	// TODO:
+	// if (m_BVS) value |= 0x10;
+	return value;
+}
+
+uint8_t dfs500_state::cpanel_reg0_r(offs_t offset)
+{
+	uint8_t data;
+	switch (offset & 0x0f)
+	{
+		case 0: // RD0
+			data = 0x00;
+			//TODO: if (m_RVD) data |= 0x02;
+			//TODO: if (m_adc->eoc_r()) data |= 0x01;
+			return data;
+		case 1: return ioport("RD1")->read();
+		case 2: return ioport("RD2")->read();
+		case 3: return ioport("RD3")->read();
+		case 4: return ioport("RD4")->read();
+		case 5: return ioport("RD5")->read();
+		case 6: return 0xFF; //TODO: RD6
+		case 7: return 0xFF; //TODO: RD7
+		case 8: return ioport("RD8")->read();
+		case 9: return ioport("RD9")->read();
+		default:
+			return 0xFF;
+	}
+}
+
+uint8_t dfs500_state::cpanel_reg2_r(offs_t offset)
+{
+	switch (offset & 0x07)
+	{
+		case 0: return ioport("RD10")->read();
+		case 1: return ioport("RD11")->read();
+		case 2: return ioport("RD12")->read();
+		default:
+			return 0xFF;
+	}
+}
+
+void dfs500_state::cpanel_reg0_w(offs_t offset, uint8_t data)
+{
+	static const uint8_t ls247_map[16] =
```

Comment:
I believe MAME already has output helpers for 7-segment displays. This function can be simplified.

Reply:
I'd be happy to learn more about it and adopt it. Do you know of an example in some other driver?

Reply:
To my surprise, we actually don't have output helpers for 7-segment displays hooked up to standard 74-series parts. Please ignore that bit of feedback, then. :)